Adds utility function…

tribeiro opened the pull request:
…to transform from spherical (ra,dec) to cartesian coordinates system and to a grid id using a spatial tree. This algorithm is used on some places on sims_maf and on sims_featureScheduler. It would be good to have it at a higher level package.

danielsf commented on python/lsst/sims/utils/treeUtils.py:
```diff
@@ -0,0 +1,72 @@
+__all__ = ['_treexyz', '_rad_length', '_buildTree']
+
+
+def _treexyz(ra, dec):
```
There is a method `cartesianFromSpherical` here

https://github.com/lsst/sims_utils/blob/master/python/lsst/sims/utils/CoordinateTransformations.py#L167

Can you use that?

danielsf commented on python/lsst/sims/utils/treeUtils.py:
```diff
@@ -0,0 +1,72 @@
+    return result
+
+
+def _buildTree(ra, dec, leafsize=100):
```
Do you mind adding a unit test that just runs `_buildTree`, in case the scipy API ever changes out from under us?

rhiannonlynne commented:
I will also request "private" and "public" versions of the methods which keep a consistent radians vs degrees interface (private = radians, public = degrees).

danielsf commented:
Lynne raises an interesting point.   `cartesianFromSpherical` (and its inverse `sphericalFromCartesian`) are old enough to predate our "`_methodName()` means radians,`methodName()` means degrees convention".  We have to keep `cartesianFromSpherical` around as it is because too much other code assumes it is there.  We could create new `xyz_from_ra_dec`, `_xyz_from_ra_dec`, `_ra_dec_from_xyz`, and `ra_dec_from_xyz` methods that call the existing methods under the hood but make the correct conversions between degrees and radians to preserver our convention.

rhiannonlynne commented:
I like those name suggestions too, Scott. They are super clear!

tribeiro commented:
Thanks for pointing this out @danielsf ... These are great suggestions indeed. I'll start working on these soon. SW review and JTM happening this week here in Tucson 🌵 ...

tribeiro commented:
Is it ok if the new `xyz_from_ra_dec`, _`xyz_from_ra_dec`, `_ra_dec_from_xyz`, and `ra_dec_from_xyz` functions together with `_rad_length` (renamed to be `xyz_angular_radius` and `_xyz_angular_radius`) are stored on https://github.com/lsst/sims_utils/blob/master/python/lsst/sims/utils/CoordinateTransformations.py ?

danielsf commented:
Yeah.  I think that is where they should go.

tribeiro commented:
Ok, I think I covered all the suggestions on this latest commit. Let me know what you think.

tribeiro commented:
Ok, done! I wan't sure about the deprecation warnings as well. In any case, I removed them. Also added unit tests to verify the relation between the different methods.

danielsf commented:
Great.  Assuming this passes a test on Jenkins, I say go ahead and merge

(note: lsst_sims always fails on the Jenkins python2 node; we just need to pass a build on the python3 nodes)

tribeiro commented:
It passed on all platforms.... https://ci.lsst.codes/job/stack-os-matrix/27612/

danielsf commented:
@tribeiro It passed because you asked Jenkins to only build `sims_utils`.  The failure comes when you try to build `lsst_sims`, which includes all of the sims packages.  In the future, please ask Jenkins to build `lsst_sims`.  A lot of packages in `lsst_sims` depend on `sims_utils`.  If we had broken something that, for instance, `sims_catUtils` depends on, the only way to catch the failure would be to ask Jenkins to build `lsst_sims`.  I am not worried about this case but, in the future, it is better to be paranoid.

(Nothing depends on `sims_maf`, so you can probably break this rule when developing on MAF)
